File: leetcode-04/solutions/python3/1003.py
class Solution:
    def isValid(self, S: str) -> bool:
        stack = []

        for c in S:
            if c == 'c':
                n = len(stack)
                if n < 2 or stack[n - 2] != 'a' or stack[n - 1] != 'b':
                    return False
                stack.pop()
                stack.pop()
            else:
                stack.append(c)

        return len(stack) == 0

File: leetcode-04/solutions/cpp/1003.cpp
class Solution {
 public:
  bool isValid(string S) {
    vector<char> stack;

    for (char c : S) {
      if (c == 'c') {
        int n = stack.size();
        if (n < 2 || stack[n - 2] != 'a' || stack[n - 1] != 'b') return false;
        stack.pop_back();
        stack.pop_back();
      } else {
        stack.push_back(c);
      }
    }

    return stack.empty();
  }
};